include/dlaf/eigensolver/tridiag_solver/merge.h

```diff
@@ -961,135 +968,432 @@ void reduceSumScalingVector(common::Pipeline<comm::Communicator>& col_task_chain
-                                           (i_tile_l.col() - idx_loc_begin.col()) * sz_loc_tiles.rows();
-          const TileElementIndex i = dist.tileElementIndex(i_g);
-          evec_tile_arr[to_sizet(i_subm_evec_arr)](i) = T{1};
-        }
-      }
+  TileCollector tc{i_begin, i_end};
+
+  const SizeType n = problemSize(i_begin, i_end, dist);
+
+  const SizeType m_subm_el_lc = [=]() {
+    const auto i_loc_being = ij_begin_lc.row();
+    const auto i_loc_end = ij_begin_lc.row() + sz_loc_tiles.rows();
+    return dist.localElementDistanceFromLocalTile<Coord::Row>(i_loc_being, i_loc_end);
+  }();
+
+  const SizeType n_subm_el_lc = [=]() {
+    const auto i_loc_being = ij_begin_lc.col();
+    const auto i_loc_end = ij_begin_lc.col() + sz_loc_tiles.cols();
+    return dist.localElementDistanceFromLocalTile<Coord::Col>(i_loc_being, i_loc_end);
+  }();
+
+  auto bcast_evals = [i_begin, i_end,
+                      dist](const dlaf::comm::Communicator& row_comm,
+                            const std::vector<matrix::Tile<T, Device::CPU>>& eval_tiles) {
+    using dlaf::comm::internal::sendBcast_o;
+    using dlaf::comm::internal::recvBcast_o;
+
+    const comm::Index2D this_rank = dist.rankIndex();
+
+    std::vector<ex::unique_any_sender<>> comms;
+    comms.reserve(to_sizet(i_end - i_begin));
+
+    for (SizeType i = i_begin; i < i_end; ++i) {
+      const comm::IndexT_MPI evecs_tile_rank = dist.rankGlobalTile<Coord::Col>(i);
+      auto& tile = eval_tiles[to_sizet(i - i_begin)];
+
+      if (evecs_tile_rank == this_rank.col())
+        comms.emplace_back(transformMPI(sendBcast_o, ex::just(std::cref(row_comm), std::cref(tile))));
+      else
+        comms.emplace_back(transformMPI(recvBcast_o, ex::just(std::cref(row_comm), evecs_tile_rank,
+                                                              std::cref(tile))));
     }
+
+    return ex::ensure_started(ex::when_all_vector(std::move(comms)));
   };
 
-  TileCollector tc{i_begin, i_end};
+  auto allReduceInPlace_o = [](const dlaf::comm::Communicator& comm, MPI_Op reduce_op, const auto& data,
+                               MPI_Request* req) {
+    auto msg = comm::make_message(data);
+    DLAF_MPI_CHECK_ERROR(MPI_Iallreduce(MPI_IN_PLACE, msg.data(), msg.count(), msg.mpi_type(), reduce_op,
+                                        comm, req));
+  };
 
-  auto sender =
-      ex::when_all(std::forward<KSender>(k), std::forward<RhoSender>(rho),
-                   ex::when_all_vector(tc.read(d)), ex::when_all_vector(tc.read(z)),
+  const std::size_t nthreads = [size = sz_loc_tiles.cols()]() {
+    const std::size_t min_workers = 1;
+    const std::size_t available_workers = getTridiagRank1NWorkers();
+    const std::size_t ideal_workers = util::ceilDiv(to_sizet(size), to_sizet(2));
+    return std::clamp(ideal_workers, min_workers, available_workers);
+  }();
+
+  ex::start_detached(
+      ex::when_all(ex::just(std::make_shared<pika::barrier<>>(nthreads)),
```

Comment:
I realized in the other PR that these can be `make_unique` instead of `make_shared`. The data isn't copied for each thread that is processing the `bulk`, but they all point to the same data, so there's no need for a `shared_ptr`. If you want you can update that here, but I will update other places in a separate PR.